kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java

```diff
@@ -705,6 +724,52 @@ protected boolean checkAuthzGroups(String userId, String function, String entity
 			azgs = ref.getAuthzGroups(userId);
 		}
 
+		// need to know whether role swap is in effect, since we can't share the cache entry between sessions
+		// that are swapped and not swapped
+
+		String siteRef = null;
+		String roleswap = null;
+
+		// Actual code in DbAuthzGroupService will not roleswap if there's a user site ref in the list and
+		// it is acceptable. However we can't tell that without doing a database access, so be conservative
+		// and cache it separately as a role swap
+		// This code does not handle delegated access
+		if (azgs != null && userId != null && userId.equals(sessionManager().getCurrentSessionUserId())) {
```

Comment:
OK my only concern is duplication of this logic, as you know Sakai has a lot of duplicate logic so it would be ideal to have this logic exist on one location that both SecurityService and DbAuthzGroupService use. Since there is a db query involved it makes sense for it to be in the latter.

If you still feel there should be 2 versions of it even though one may be slightly simplified at least leave a comment.